src/stores/ScannerStore.js

```diff
@@ -15,57 +15,114 @@
 // along with Parity.  If not, see <http://www.gnu.org/licenses/>.
 
 // @flow
+
+import { createType } from '@polkadot/types';
 import { Container } from 'unstated';
-import { NETWORK_LIST, NetworkProtocols, EthereumNetworkKeys } from '../constants';
+
+import { NETWORK_LIST, NetworkProtocols } from '../constants';
 import { saveTx } from '../util/db';
-import { brainWalletSign, decryptData, keccak, ethSign } from '../util/native';
+import { isAscii } from '../util/message';
+import { blake2s, brainWalletSign, decryptData, keccak, ethSign, substrateSign } from '../util/native';
 import transaction from '../util/transaction';
+import { constructDataFromBytes, asciiToHex } from '../util/decoders';
 import { Account } from './AccountsStore';
 
 type TXRequest = Object;
 
 type SignedTX = {
-  txRequest: TXRequest,
+  recipient: Account,
   sender: Account,
-  recipient: Account
+  txRequest: TXRequest,
 };
 
 type ScannerState = {
-  type: 'transaction' | 'message',
-  txRequest: TXRequest | null,
+  dataToSign: string,
+  isHash: boolean,
+  isOversized: boolean,
   message: string,
-  tx: Object,
-  sender: Account,
+  multipartData: any,
   recipient: Account,
-  dataToSign: string,
-  signedData: string,
   scanErrorMsg: string,
-  signedTxList: [SignedTX]
+  sender: Account,
+  signedData: string,
+  signedTxList: [SignedTX],
+  tx: Object,
+  txRequest: TXRequest | null,
+  type: 'transaction' | 'message',
+  unsignedData: any
 };
 
 const defaultState = {
-  type: null,
   busy: false,
-  txRequest: null,
+  isHash: false,
+  isOversized: false,
+  dataToSign: '',
   message: null,
-  sender: null,
+  multipartData: {},
   recipient: null,
-  tx: '',
-  dataToSign: '',
+  scanErrorMsg: '',
+  sender: null,
   signedData: '',
-  scanErrorMsg: ''
+  tx: '',
+  txRequest: null,
+  type: null,
+  unsignedData: {}
 };
 
 export default class ScannerStore extends Container<ScannerState> {
   state = defaultState;
 
-  async setData(data, accountsStore) {
-    console.log(data);
-    switch (data.action) {
+  async setUnsigned(data) {
+    this.setState({
+      unsignedData: JSON.parse(data)
+    });
+  }
+
+  async setParsedData(strippedData, accountsStore) {
+    const parsedData = await constructDataFromBytes(strippedData);
+    
+    if (parsedData.isMultipart) {
+      this.setPartData(parseData.frame, parsedData.frameCount, parseData.partData, accountsStore);
+      return;
+    }
+
+    this.setState({
+      unsignedData: parsedData
+    });
+  }
+
+  async setPartData(frame, frameCount, partData, accountsStore) {
+    if (partData[0] === new Uint8Array([0x00]) || partData[0] === new Uint8Array([0x7B])) {
+      // part_data for frame 0 MUST NOT begin with byte 00 or byte 7B.
```

Comment:
why not zero, does zero means [Multipart Payload](https://github.com/maciejhirsz/uos#multipart-payload)

Reply:
Yes, if 0th byte for part data is 0, then the part data will itself be treated as multipart and split it again, which is not the intended behavior. https://github.com/maciejhirsz/uos#multipart-payload